### DISCUSSION TITLE
filter bonus exercise

### DISCUSSION THREAD

**Discussion by BenjaminGolden** [nss-day-cohort-47/C47-Discussions#75] [Help Me] [answered]
B.J. Golden

I am working on the bonus question in the filter array exercises and I'm looking for some direction for how I can match the specific properties that are passed in as parameters.  

```export const getStudentsByLanguages = ([languages]) => {
    const studentsThatMatchLanguages = students.filter(student => student.languages.some(studentLanguage => studentLanguage === languages))
        return studentsThatMatchLanguages;
}
```

This code will return an array of students that have the specific languages passed in as parameters, but it will not return the students who specifically match the given parameters. Ex. If Javascript and C# are called using ```getStudentsByLanguages(["Javascript", "C#"])``` it will return a student that has Javascript, but not C# and vice versa. I assume this is because .some evaluates to true or false and that when the function runs it is matching that parameter which evaluates to true. I'm not sure how to be more specific using the given syntax that I have in order to return the students who match both specific languages passed in as parameters. 

I've tried researching and using .every, .some, and .find. .some seems to be the best choice in this case but I'm missing something and I can't seem to figure out what that is. 

https://github.com/BenjaminGolden/Array-Exercises/blob/main/src/filterExercises.js

line 96-100

thanks,

B.J.

**Comment by BryanNilsen** [comment]
- make sure you're naming the function properly according to the instructions
- think about what you're trying to do... you want to only return the students who have a match on each language in the array that's being passed in.
- this leads me to suggest starting with a filter on the students array - and then as we iterate the students, we want to check and see if **every** language being passed into the function has a match within the current student's languages array.
- hope that makes sense. this is tough!

**Comment by Chadwell-Clark** [comment]
I was having trouble with this one and ended up with  vanilla for loops to iterate through students and then conditionally using if statements to  go through the languages passed in and compare them to the array of student languages. I got it down to one for loop while using every and includes to see if the language was in the student languages. If it was i pushed it into an array , if not I did not. I then refactored it to use a forEach and then refactored further and swapped a filter in for the forEach. I started with a huge script that was greatly reduced when refactored. In the end I did not need to push it into an array as filter did that for me. Breaking it down like Bryan is saying above was the only way for me to think about it properly. I wrote it at least 6 times before just going back to basics and working it from the ground up. The process helped me learn every and includes. It was tough.  I made a new JS file with the dataset and worked on the exercise there so I could run node and see my console.logs. Alternatively you could make a simple index.html so you could serve it up and work on it in the browser with the console and dev tools

**Comment by BryanNilsen** [comment, marked as answer]
Let me preface this by saying this was challenging to work out the logic.
Here is my solution: 
```
export const getStudentsByLanguages = (languages) => students.filter(student => languages.every(lang => student.languages.includes(lang)))
```
FWIW - I don't expect anyone to have done it this way. This is just demonstrating the power of array methods.

You may have to stare at this for a while to see what's going on. It might not make any sense even after staring at it.
I'd love to see other solutions, if anyone wants to post them!!

**Reply by BenjaminGolden** [reply]
I'm going to try to articulate what is going on in your code to see if I'm understanding this correctly. 
We are filtering through the students array, then for each student, we are looking at the languages which were passed in as parameters. Then we are checking if every language of the parameters is contained in the languages of the current student we are filtering over. Is this close? I've been working with this in the exercise and it still won't pass the test. ```getStudentsByLanguages``` returns undefined.

**Reply by BryanNilsen** [reply]
Your articulation is correct! Great work, BJ!!

As far as your code goes, are you returning the filtered array?
Post your code so we can take a look.

**Comment by Chadwell-Clark** [comment]
Here is my code about halfway through my refactoring:
```
export const getStudentsByLanguages = (languages) => {
  let studentsAll = useStudents();
  let stuWithLang = [];
  studentsAll.forEach((student) => {
    if (languages.every((lang) => student.languages.includes(lang))) {
      console.log("every lang found in", student);
      stuWithLang.push(student);
    } else {
      console.log("NOT every lang found in", student);
    }
  });
  return stuWithLang;
};
```
at this point I tried substituting filter for the forEach and then ended up with something like Bryan's. Bryan helped me get there. Chris had an interesting solution and Bryson had one similar to Bryan's.

**Reply by BryanNilsen** [reply]
I like seeing your thought process in this code!

**Reply by BenjaminGolden** [reply]
In the if statement, ```languages.every``` is calling the ```(languages)``` parameter of ```getStudentsByLanguages``` ? Is ```(lang)``` just a taco that you are storing the language found in the parameter?  I'm still struggling with how this one is working. Thank you for sharing your process!

**Reply by Chadwell-Clark** [reply]
Yes` lang` is a taco for the current element  of the `languages` array that is being evaluated with the conditional (does the student.languages array include the current` lang `element)
Both every() and includes() return booleans so the if statement will be true only when the student languages array includes() is true for **every** `lang` in the languages array passed into `getStudentsByLanguages()`
The if statement is pushing the student object into a new array if the student object includes all of the `languages` in the languages array passed into the `getStudentsByLanguages`. If anything returns false the forEach goes to the next student. You end up with an array of just students with all the languages.
